In [1]:
# Imports and pip installations (if needed)
import sklearn
import pandas as pd
import sklearn.model_selection
import sklearn.linear_model
import sklearn.svm
import sklearn.neural_network
import sklearn.neighbors

# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely or locally)
ds = pd.read_csv("iris.csv")
# shuffling the data
ds = ds.sample(frac = 1)
# Output the first 15 rows of the data
print(ds.head(n=15))
# Display a summary of the table information (number of datapoints, etc.)
ds.info()
print(ds.shape)
print(ds.describe())


     sepal_length  sepal_width  petal_length  petal_width     species
22            4.6          3.6           1.0          0.2      setosa
133           6.3          2.8           5.1          1.5   virginica
97            6.2          2.9           4.3          1.3  versicolor
40            5.0          3.5           1.3          0.3      setosa
98            5.1          2.5           3.0          1.1  versicolor
89            5.5          2.5           4.0          1.3  versicolor
76            6.8          2.8           4.8          1.4  versicolor
119           6.0          2.2           5.0          1.5   virginica
6             4.6          3.4           1.4          0.3      setosa
5             5.4          3.9           1.7          0.4      setosa
42            4.4          3.2           1.3          0.2      setosa
124           6.7          3.3           5.7          2.1   virginica
110           6.5          3.2           5.1          2.0   virginica
0             5.1   

## About the dataset
The data represents the dimensions of the petals and sepals (modified leaves that resemble petals in orchids) of various photos of orchids as the features and labels each with the correct species of orchid represented in the photo. No null values are present, and there are 150 such processed representations of photos in the dataset. Before being split, the dataset thus has 150 rows and 5 columns. 

# Part 2: Split the dataset into train and test

In [3]:
# Take the dataset and split it into our features (X) and label (y)

X = ds.loc[:, ["sepal_length", "sepal_width", "petal_length", "petal_width"]]
y = ds["species"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
Xtrain, Xtest, ytrain, ytest = sklearn.model_selection.train_test_split(X, y, test_size = 0.1)

# Part 3: Logistic Regression

In [ ]:
# i. Use sklearn to train a LogisticRegression model on the training set
logreg = sklearn.linear_model.LogisticRegression(max_iter=200)
logreg.fit(Xtrain, ytrain)
# ii. For a sample datapoint, predict the probabilities for each possible class
sampledp, samplecorrect = Xtest.iloc[[5]], ytest.iloc[[5]]
sampleresult = logreg.predict(sampledp)
print("The model predicted that the image was of a ", sampleresult[0], " iris. The image was of a ", 
      samplecorrect.iat[0], "iris. ")
# iii. Report on the score for Logistic regression model, what does the score measure?
logregscore = logreg.score(Xtest, ytest)
print("The score for the logistic regression model is ", logregscore)
# iv. Extract the coefficents and intercepts for the boundary line(s)
print("Boundary line coefficients: ", logreg.coef_)
print("Boundary line intercepts: ", logreg.intercept_)

The model predicted that the image was of a  virginica  iris. The image was of a  virginica iris. 
The score for the logistic regression model is  0.9333333333333333
Boundary line coefficients:  [[-0.44359433  0.8769705  -2.45261969 -1.04247899]
 [ 0.34580314 -0.31793023 -0.18792822 -0.79657568]
 [ 0.09779119 -0.55904027  2.64054791  1.83905467]]
Boundary line intercepts:  [ 10.05559915   3.02926568 -13.08486483]


The score for the model was 0.93333, indicating that the model was 93.333% accurate at predicting an orchid's species given its petal and sepal dimensions even when that orchid had not been trained on. The boundary coefficients and intercepts are listed above.

# Part 4: Support Vector Machine

In [5]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svc = sklearn.svm.SVC()
svc.fit(Xtrain, ytrain)
# ii. For a sample datapoint, predict the probabilities for each possible class
sampledp, samplecorrect = Xtest.iloc[[11]], ytest.iloc[[11]]
sampleresult = svc.predict(sampledp)
print("The model predicted that the image was of a ", sampleresult[0], " iris. The image was of a ", 
      samplecorrect.iat[0], "iris. ")
# iii. Report on the score for the SVM, what does the score measure?
# TODO: answer the questions
svcscore = svc.score(Xtest, ytest)
print(svcscore)

The model predicted that the image was of a  virginica  iris. The image was of a  virginica iris. 
0.9333333333333333


The SVM got a score of 0.93333, indicating that when presented with new data from the testing set, the model was accurate 93.333% of the time in identifying the correct species.

# Part 5: Neural Network

In [12]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
nn = sklearn.neural_network.MLPClassifier()
nn.fit(Xtrain, ytrain)
# ii. For a sample datapoint, predict the probabilities f6or each possible class
sampledp, samplecorrect = Xtest.iloc[[14]], ytest.iloc[[14]]
sampleresult = svc.predict(sampledp)
print("The model predicted that the image was of a ", sampleresult[0], " iris. The image was of a ", 
      samplecorrect.iat[0], "iris. ")
# iii. Report on the score for the Neural Network, what does the score measure?
#TODO: Answer questions
nnscore = nn.score(Xtest, ytest)
print(nnscore)
# iv: Experiment with different options for the neural network, report on your best configuration
nn2 = sklearn.neural_network.MLPClassifier(activation = 'relu', learning_rate='adaptive', max_iter=3000)
nn2.fit(Xtrain, ytrain)
sampledp, samplecorrect = Xtest.iloc[[14]], ytest.iloc[[14]]
sampleresult = svc.predict(sampledp)
print("The model predicted that the image was of a ", sampleresult[0], " iris. The image was of a ", 
      samplecorrect.iat[0], "iris. ")
nn2score = nn2.score(Xtest, ytest)
print(nn2score)

/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


The model predicted that the image was of a  setosa  iris. The image was of a  setosa iris. 
0.9333333333333333
The model predicted that the image was of a  setosa  iris. The image was of a  setosa iris. 
1.0


The score for the first model with default settings was 0.933333, which indicated that the model was 93.33% accurate with its predictions in the testing dataset. In contrast, when increasing the maximum number of iterations, changing the activation function to ReLU, and setting an adaptive learning rate so that steps get smaller as convergence is neared, the score reached 1.0, indicating 100% predictive accuracy with the testing dataset.

# Part 6: K-Nearest Neighbors

In [7]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn = sklearn.neighbors.KNeighborsClassifier()
knn.fit(Xtrain, ytrain)
# ii. For a sample datapoint, predict the probabilities for each possible class
sampledp, samplecorrect = Xtest.iloc[[9]], ytest.iloc[[9]]
sampleresult = svc.predict(sampledp)
print("The model predicted that the image was of a ", sampleresult[0], " iris. The image was of a ", 
      samplecorrect.iat[0], "iris. ")
# iii. Report on the score for kNN, what does the score measure?
knnscore = knn.score(Xtest, ytest)
print(knnscore)
#TODO: answer these questions

The model predicted that the image was of a  versicolor  iris. The image was of a  versicolor iris. 
0.9333333333333333


The K nearest neighbors model had a score of 0.933333, indicating that it was correct 93.3333% of the time in predicting the species of an orchid given petal and sepal dimensions with the new data in the testing dataset.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

I was surprised by the consistency in results between all of the models except for the neural networks, all staying at around 93% accuracy when faced with new data. It is unsurprising to me that the neural network performed the best, given that it allows for more sophisticated relationships between parts of the data. Since orchids require a high level of expertise among botanists to distinguish, it makes sense that a more complex relationship would be needed. Perhaps some of these other models could have performed better given more features representing polynomial representations of the other measurements or additional features representing color or the curvature of the petals and sepals, though these measurements may be hard to control in the dataset; for instance, if color were added as a feature that was extracted from the photos we would have to worry that certain species were photographed in one particular greenhouse with a different tint to the light.